### PR TITLE
Remove vendored flax layers

### PR DESCRIPTION
**Pull request by PhilipVinc**
https://github.com/google/flax/pull/2058 is going to land soon.

Once they tag a new version, I will bump NetKet's minimum flax version to match and would like to merge this PR.
This comes with the deletion of 400 lines of code (and we add 150 that aree mainly deprecation warnings...)

This PR removes our own implementation of Dense and Conv layers, deprecating `nk.nn.Dense/Conv` and instructing users to use `flax.linen.Dense` and `flax.linen.Conv`.

Post-flax/#2058 their layers will have two attributes: `dtype` and `param_dtype`. 
 - `param_dtype` specifies the dtype of the parameters in the layer (and is equivalent to netket's current `dtype` attribute`)     
 -  `dtype` is an optional type, `None` by default, and specifies the precision used to compute the layer. This will only work on accelerators (GPU/TPU) . 

Most likely, netket's users won't need `dtype` and only `param_dtype`.

This will only affect those that define their own architectures, and the error message is quite clear. The old syntax will still be supported for the time being with a (very loud) deprecation warning.

### CODE REVIEW

**Comment by PhilipVinc** [comment]
As flax will transit to use `param_dtype`, we could consider to change RBM/GCNN/etc `dtype` to `param_dtype` for consistency.

File: docs/tutorials/gs-ising.ipynb [hunk 1]
```diff
@@ -1172,9 +1172,9 @@
     "        # the shape will now be (batches, 1, Nsites)\n",
     "        x = x.reshape(-1, 1, x.shape[-1])\n",
     "        \n",
-    "        x = nknn.DenseSymm(symmetries=graph.translation_group(),\n",
-    "                           features=self.alpha,\n",
-    "                           kernel_init=nn.initializers.normal(stddev=0.01))(x)\n",
+    "        x = nn.DenseSymm(symmetries=graph.translation_group(),\n",
+    "                         features=self.alpha,\n",
+    "                         kernel_init=nn.initializers.normal(stddev=0.01))(x)\n",
```
Comment (femtobit) [inline docs/tutorials/gs-ising.ipynb:1177]: I don't think `DenseSymm` should be changed to `flax`.

Comment (PhilipVinc) [reply docs/tutorials/gs-ising.ipynb:1177]: ops

**Comment by PhilipVinc** [comment]
Changes where merged upstream and now flax 0.5 correctly supports complex numbers! 

This PR is ready to go, there's only a bunch of issues with the `test-oldest-deps` because pip is very stupid...

**Comment by PhilipVinc** [comment]
Technically the only dependency that we had to bump was flax, but then the script that tests oldest dependencies broke because he cannot handle well weird situations. Sso I also bumped a few others:
from
```python
    "numpy~=1.19",
     "jax>=0.3.2, <0.4",
     "jaxlib>=0.3",
```
to
```
    "numpy~=1.19",
     "jax>=0.3.2, <0.4",
     "jaxlib>=0.3",
```

`Flax 0.5` requires `jax 0.3.2` so I bumped it as well. `jax 0.3` requires numpy `1.19` and `jaxlib 0.3` so I bumped those as well.
All in all, we're just propagating their requirements, so it does not further restrict available versions.

**Comment by PhilipVinc** [comment]
Tests pass and this is ready to go

**Review by gcarleo — APPROVED** [review]

**Pull request merged**